odoo/tools/xml_utils.py

```diff
@@ -139,3 +145,133 @@ def leaf_iter(parent_node, node, level):
+    Filtering which file will be saved can be done by providing a list of `xsd_names`
+
+    The XSD files content can be modified by providing the `modify_xsd_content` function as argument.
+    Typically, this is used when XSD files depend on each other (with the schemaLocation attribute),
+    but it can be used for any purpose.
+
+    :param odoo.api.Environment env: environment of calling module
+    :param str url: URL of XSD file/ZIP archive
+    :param str xsd_name: the name given to the XSD attachment
+    :param bool force_reload: if True, reload the attachment from URL, even if it is already cached
+    :param int request_max_timeout: maximum time (in seconds) before the request times out
+    :param str xsd_name_prefix: if provided, will be added as a prefix to every XSD file name
+    :param list | str xsd_names: if provided, will only save the XSD files with these names
+    :param func modify_xsd_content: function that takes the xsd content as argument and returns a modified version of it
+    :rtype: odoo.api.ir.attachment | bool
+    :return: the main attachment or False if an error occurred (see warning logs)
+    """
+    if not url.endswith(('.xsd', '.zip')):
+        _logger.warning("The given URL (%s) needs to lead to an XSD file or a ZIP archive", url)
+        return False
+
+    is_zip = url.endswith('.zip')
+
+    fetched_attachment = env['ir.attachment'].search([('name', '=', xsd_name)])
+    if fetched_attachment:
+        if not force_reload:
+            _logger.info("Retrieved attachment from database, with name: %s", fetched_attachment.name)
+            return fetched_attachment
+        _logger.info("Found the attachment with name %s in database, but forcing the reloading.", fetched_attachment.name)
+
+    try:
+        _logger.info("Fetching file/archive from given URL: %s", url)
+        response = requests.get(url, timeout=request_max_timeout)
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as error:
+        _logger.warning('HTTP error: %s with the given URL: %s', error, url)
+        return False
+    except requests.exceptions.ConnectionError as error:
+        _logger.warning('Connection error: %s with the given URL: %s', error, url)
+        return False
+    except requests.exceptions.Timeout as error:
+        _logger.warning('Request timeout: %s with the given URL: %s', error, url)
+        return False
+
+    content = response.content
+    if modify_xsd_content and not is_zip:
+        content = modify_xsd_content(content)
+
+    if fetched_attachment:
+        _logger.info("Updating the content of ir.attachment with name: %s", xsd_name)
+        fetched_attachment.update({
+            'raw': content
+        })
+        return fetched_attachment
+
+    _logger.info("Saving XSD file as ir.attachment, with name: %s", xsd_name)
+    main_attachment = env['ir.attachment'].create({
+        'name': xsd_name,
```

Comment:
variable's name is a bit confusing as it's not necesarily an xsd's name (from what I understand if the file is a zip then it's the zip's name).